File: 2021/src/day1.fs
module day1

type foldState = { inc: int; cur: int }

let input =
    "./input/day1input.txt"
    |> System.IO.File.ReadAllLines
    |> Array.map System.Int32.Parse

let p1 =
    input
    |> Array.fold
        (fun acc value ->
            let nextInc =
                match value with
                | x when x > acc.cur -> acc.inc + 1
                | _ -> acc.cur

            { inc = nextInc; cur = value })
        { inc = 0; cur = System.Int32.MaxValue }

printfn "%i" p1.inc

let p2 =
    input
    |> Array.indexed
    |> Array.fold
        (fun acc value ->
            let index = fst value

            let sum =
                input
                |> Array.append [| 0; 0 |]
                |> fun x -> Array.sub x index 3
                |> Array.sum

            let nextInc =
                match sum with
                | x when x > acc.cur -> acc.inc + 1
                | _ -> acc.cur

            { inc = nextInc; cur = sum })
        { inc = 0; cur = System.Int32.MaxValue }

printfn "%i" p2.inc

File: 2021/src/day1.clj
(ns day4 (:require [clojure.string :refer [split-lines]]))

(def input (-> "./input/day1input.txt"
               clojure.core/slurp
               split-lines
               (as-> t (map #(. Integer parseInt %) t))
               (as-> t (into [] t))))

(def part1
  (reduce
   (fn [{:keys [count cur]} val]
     (let [nextCount (if (> val cur) (inc count) count)]
       {:cur val :count nextCount}))
   {:count 0 :cur Integer/MAX_VALUE} input))

(def part2
  (reduce-kv
   (fn [{:keys [count cur]} index _]
     (let [current (apply + (subvec (into input [0 0]) index (+ 3 index)))
           nextCount (if (> current cur) (inc count) count)]
       {:cur current :count nextCount}))
   {:count 0 :cur Integer/MAX_VALUE} input))
